File: 383-ransom-note/ransom-note.py
class Solution:
    def canConstruct(self, ransomNote: str, magazine: str) -> bool:
        mp = defaultdict(int)
        for c in magazine:
            mp[c] += 1
        
        for c in ransomNote:
            if c not in mp:
                return False
            mp[c] -= 1
            if mp[c] == 0:
                del mp[c]
        return True

File: 383-ransom-note/ransom-note.cpp
class Solution {
public:
    bool canConstruct(string ransomNote, string magazine) {
        if (ransomNote.size() > magazine.size()) {
            return false;
        }

        int count[26] = {0};
        for (char c : magazine) {
            count[c - 'a']++;
        }

        for (char c : ransomNote) {
            count[c - 'a']--;
            if (count[c - 'a'] < 0) {
                return false;
            }
        }
        return true;
    }
};